builtin/aws/ecs/platform.go

```diff
@@ -1045,75 +1049,31 @@ func (p *Platform) resourceAlbListenerCreate(
 
 	elbsrv := elbv2.New(sess)
 
-	var listener *elbv2.Listener
-
-	if albConfig != nil && albConfig.ListenerARN != "" {
-		s.Update("Describing requested ALB listener (ARN: %s)", albConfig.ListenerARN)
-
-		state.Managed = false
-
-		out, err := elbsrv.DescribeListenersWithContext(ctx, &elbv2.DescribeListenersInput{
-			ListenerArns: []*string{aws.String(albConfig.ListenerARN)},
-		})
-		if err != nil {
-			return status.Errorf(codes.Internal, "failed to describe requested listener ARN %q: %s", albConfig.ListenerARN, err)
-		}
-
-		listener = out.Listeners[0]
-		s.Update("Using configured ALB Listener: %s (load-balancer: %s)",
-			*listener.ListenerArn, *listener.LoadBalancerArn)
-	} else {
-		state.Managed = true
-
-		if alb == nil || alb.Arn == "" {
-			return status.Errorf(codes.InvalidArgument, "cannot create ALB listener - no existing ALB defined.")
-		}
-
-		s.Update("No ALB listener specified - looking for listeners for ALB %q", alb.Name)
-		listeners, err := elbsrv.DescribeListenersWithContext(ctx, &elbv2.DescribeListenersInput{
-			LoadBalancerArn: &alb.Arn,
-		})
-		if err != nil {
-			return status.Errorf(codes.Internal, "failed to describe listeners for alb (ARN %q): %s", alb.Arn, err)
-		}
-
-		if len(listeners.Listeners) > 0 {
-			listener = listeners.Listeners[0]
-			s.Update("Using existing ALB Listener (ARN: %q)", *listener.ListenerArn)
-		} else {
-			s.Update("Creating new ALB Listener")
-			newListener = true
-
-			log.Info("load-balancer defined", "dns-name", alb.DnsName)
-
-			tgs[0].Weight = aws.Int64(100)
-			lo, err := elbsrv.CreateListenerWithContext(ctx, &elbv2.CreateListenerInput{
-				LoadBalancerArn: &alb.Arn,
-				Port:            aws.Int64(int64(externalIngressPort)),
-				Protocol:        aws.String(protocol),
-				Certificates:    certs,
-				DefaultActions: []*elbv2.Action{
-					{
-						ForwardConfig: &elbv2.ForwardActionConfig{
-							TargetGroups: tgs,
-						},
-						Type: aws.String("forward"),
-					},
-				},
-			})
-			if err != nil {
-				return status.Errorf(codes.Internal, "failed to create listener: %s", err)
-			}
+	if alb == nil || alb.Arn == "" {
```

Comment:
❓ I see that this if block already existed, but how is it enforced that the ALB is created before the listener is created? In the resource manager I can see that the ALB listener is sequentially after the ALB itself - is that what ensures that it is done first?

Reply:
Argmapper magic! This `resourceAlbListenerCreate` function takes an `*Resource_Alb` as input, and goes into the argmapper soup along with the `resourceAlbCreate` function that provides it. Argmapper figures out the right order to invoke the whole string of resource functions.